### ISSUE TITLE
Unable to use R kernel in 0.7.0

### ISSUE THREAD

**Issue by mconigliaro** [jupyterhub/jupyterhub#899]
After creating an R environment on the command line (#897), it appears under the Conda tab, but clicking on it returns a "Not Found" error. Also, my R kernels don't show up in any of the notebook kernel lists (#898 didn't help in this case).

**Comment by willingc** [comment]
Hi @mconigliaro.

Sorry for the delayed response. @minrk and I have been traveling. I see issues #897 and #898 are related to this one. Let's consolidate the discussion here to help resolve getting R kernel behave as you expect.

First, I want to make sure that we understand a bit more about your deployment.

- Was JupyterHub installed using conda/Anaconda as a new deployment or did you upgrade an existing deployment?

- Which spawner are you using?

- What is the output of running `jupyter troubleshoot` from the command line?

- What is the output of running `jupyter kernelspec list`?

Thanks!

**Comment by mconigliaro** [comment]
- This is a brand new deployment. After installing Conda 4.2.0, we ran `pip install jupyterhub` 
- [SudoSpawner](https://github.com/jupyterhub/sudospawner)
- `jupyter troubleshoot`:
```
You are using pip version 8.1.2, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
$PATH:
	/opt/anaconda3/bin
	/opt/anaconda3/bin
	/usr/local/sbin
	/usr/local/bin
	/usr/sbin
	/usr/bin
	/sbin
	/bin
	/usr/lib/jvm/java-8-oracle/bin
	/usr/lib/jvm/java-8-oracle/db/bin
	/usr/lib/jvm/java-8-oracle/jre/bin

sys.path:
	/opt/anaconda3/bin
	/opt/anaconda3/lib/python35.zip
	/opt/anaconda3/lib/python3.5
	/opt/anaconda3/lib/python3.5/plat-linux
	/opt/anaconda3/lib/python3.5/lib-dynload
	/opt/anaconda3/lib/python3.5/site-packages
	/opt/anaconda3/lib/python3.5/site-packages/Sphinx-1.4.6-py3.5.egg
	/opt/anaconda3/lib/python3.5/site-packages/setuptools-27.2.0-py3.5.egg

sys.executable:
	/opt/anaconda3/bin/python

sys.version:
	3.5.2 |Anaconda 4.2.0 (64-bit)| (default, Jul  2 2016, 17:53:06)
	[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]

platform.platform():
	Linux-3.13.0-74-generic-x86_64-with-debian-jessie-sid

which -a jupyter:
	/opt/anaconda3/bin/jupyter
	/opt/anaconda3/bin/jupyter

pip list:
	alabaster (0.7.9)
	alembic (0.8.9)
	anaconda-clean (1.0)
	anaconda-client (1.5.1)
	anaconda-navigator (1.3.1)
	argcomplete (1.0.0)
	astroid (1.4.7)
	astropy (1.2.1)
	Babel (2.3.4)
	backports.shutil-get-terminal-size (1.0.0)
	beautifulsoup4 (4.5.1)
	bitarray (0.8.1)
	blaze (0.10.1)
	bokeh (0.12.2)
	boto (2.42.0)
	Bottleneck (1.1.0)
	cffi (1.7.0)
	chest (0.2.3)
	click (6.6)
	cloudpickle (0.2.1)
	clyent (1.2.2)
	colorama (0.3.7)
	conda (4.2.9)
	conda-build (2.0.2)
	configobj (5.0.6)
	contextlib2 (0.5.3)
	cryptography (1.5)
	cycler (0.10.0)
	Cython (0.24.1)
	cytoolz (0.8.0)
	dask (0.11.0)
	datashape (0.5.2)
	decorator (4.0.10)
	dill (0.2.5)
	docutils (0.12)
	dynd (0.7.3.dev1)
	et-xmlfile (1.0.1)
	fastcache (1.0.2)
	filelock (2.0.6)
	Flask (0.11.1)
	Flask-Cors (2.1.2)
	gevent (1.1.2)
	greenlet (0.4.10)
	h5py (2.6.0)
	HeapDict (1.0.0)
	idna (2.1)
	imagesize (0.7.1)
	ipykernel (4.5.0)
	ipython (5.1.0)
	ipython-genutils (0.1.0)
	ipywidgets (5.2.2)
	itsdangerous (0.24)
	jdcal (1.2)
	jedi (0.9.0)
	Jinja2 (2.8)
	jsonschema (2.5.1)
	jupyter (1.0.0)
	jupyter-client (4.4.0)
	jupyter-console (5.0.0)
	jupyter-core (4.2.0)
	jupyterhub (0.7.0)
	jupyterhub-ldapauthenticator (1.1)
	lazy-object-proxy (1.2.1)
	ldap3 (2.1.1)
	llvmlite (0.13.0)
	locket (0.2.0)
	lxml (3.6.4)
	Mako (1.0.6)
	MarkupSafe (0.23)
	matplotlib (1.5.3)
	mistune (0.7.3)
	mpmath (0.19)
	multipledispatch (0.4.8)
	nb-anacondacloud (1.2.0)
	nb-conda (2.0.0)
	nb-conda-kernels (2.0.0)
	nbconvert (4.2.0)
	nbformat (4.1.0)
	nbpresent (3.0.2)
	networkx (1.11)
	nltk (3.2.1)
	nose (1.3.7)
	notebook (4.2.3)
	numba (0.28.1)
	numexpr (2.6.1)
	numpy (1.11.1)
	odo (0.5.0)
	openpyxl (2.3.2)
	pamela (0.3.0)
	pandas (0.18.1)
	partd (0.3.6)
	path.py (0.0.0)
	pathlib2 (2.1.0)
	patsy (0.4.1)
	pep8 (1.7.0)
	pexpect (4.0.1)
	pickleshare (0.7.4)
	Pillow (3.3.1)
	pip (8.1.2)
	pkginfo (1.3.2)
	ply (3.9)
	prompt-toolkit (1.0.3)
	psutil (4.3.1)
	ptyprocess (0.5.1)
	py (1.4.31)
	pyasn1 (0.1.9)
	pycosat (0.6.1)
	pycparser (2.14)
	pycrypto (2.6.1)
	pycurl (7.43.0)
	pyflakes (1.3.0)
	Pygments (2.1.3)
	pylint (1.5.4)
	pyOpenSSL (16.0.0)
	pyparsing (2.1.4)
	pytest (2.9.2)
	python-dateutil (2.5.3)
	python-editor (1.0.3)
	pytz (2016.6.1)
	PyYAML (3.12)
	pyzmq (15.4.0)
	QtAwesome (0.3.3)
	qtconsole (4.2.1)
	QtPy (1.1.2)
	redis (2.10.5)
	requests (2.11.1)
	rope-py3k (0.9.4.post1)
	ruamel-yaml (-VERSION)
	scikit-image (0.12.3)
	scikit-learn (0.17.1)
	scipy (0.18.1)
	setuptools (27.2.0)
	simplegeneric (0.8.1)
	singledispatch (3.4.0.3)
	six (1.10.0)
	snowballstemmer (1.2.1)
	sockjs-tornado (1.0.3)
	Sphinx (1.4.6)
	spyder (3.0.0)
	SQLAlchemy (1.0.13)
	statsmodels (0.6.1)
	sudospawner (0.3.0)
	sympy (1.0)
	tables (3.2.3.1)
	terminado (0.6)
	toolz (0.8.0)
	tornado (4.4.1)
	traitlets (4.3.0)
	unicodecsv (0.14.1)
	wcwidth (0.1.7)
	Werkzeug (0.11.11)
	wheel (0.29.0)
	widgetsnbextension (1.2.6)
	wrapt (1.10.6)
	xlrd (1.0.0)
	XlsxWriter (0.9.3)
	xlwt (1.1.2)

conda list:
	# packages in environment at /opt/anaconda3:
	#
	_license                  1.1                      py35_1
	_nb_ext_conf              0.3.0                    py35_0
	alabaster                 0.7.9                    py35_0
	alembic                   0.8.9                     <pip>
	anaconda                  4.2.0               np111py35_0
	anaconda-clean            1.0.0                    py35_0
	anaconda-client           1.5.1                    py35_0
	anaconda-navigator        1.3.1                    py35_0
	argcomplete               1.0.0                    py35_1
	astroid                   1.4.7                    py35_0
	astropy                   1.2.1               np111py35_0
	babel                     2.3.4                    py35_0
	backports                 1.0                      py35_0
	beautifulsoup4            4.5.1                    py35_0
	bitarray                  0.8.1                    py35_0
	blaze                     0.10.1                   py35_0
	bokeh                     0.12.2                   py35_0
	boto                      2.42.0                   py35_0
	bottleneck                1.1.0               np111py35_0
	cairo                     1.12.18                       6
	cffi                      1.7.0                    py35_0
	chest                     0.2.3                    py35_0
	click                     6.6                      py35_0
	cloudpickle               0.2.1                    py35_0
	clyent                    1.2.2                    py35_0
	colorama                  0.3.7                    py35_0
	conda                     4.2.9                    py35_0
	conda-build               2.0.2                    py35_0
	configobj                 5.0.6                    py35_0
	contextlib2               0.5.3                    py35_0
	cryptography              1.5                      py35_0
	curl                      7.49.0                        1
	cycler                    0.10.0                   py35_0
	cython                    0.24.1                   py35_0
	cytoolz                   0.8.0                    py35_0
	dask                      0.11.0                   py35_0
	datashape                 0.5.2                    py35_0
	dbus                      1.10.10                       0
	decorator                 4.0.10                   py35_0
	dill                      0.2.5                    py35_0
	docutils                  0.12                     py35_2
	dynd-python               0.7.2                    py35_0
	entrypoints               0.2.2                    py35_0
	et_xmlfile                1.0.1                    py35_0
	expat                     2.1.0                         0
	fastcache                 1.0.2                    py35_1
	filelock                  2.0.6                    py35_0
	flask                     0.11.1                   py35_0
	flask-cors                2.1.2                    py35_0
	fontconfig                2.11.1                        6
	freetype                  2.5.5                         1
	get_terminal_size         1.0.0                    py35_0
	gevent                    1.1.2                    py35_0
	glib                      2.43.0                        1
	greenlet                  0.4.10                   py35_0
	gst-plugins-base          1.8.0                         0
	gstreamer                 1.8.0                         0
	h5py                      2.6.0               np111py35_2
	harfbuzz                  0.9.39                        1
	hdf5                      1.8.17                        1
	heapdict                  1.0.0                    py35_1
	icu                       54.1                          0
	idna                      2.1                      py35_0
	imagesize                 0.7.1                    py35_0
	ipykernel                 4.5.0                    py35_0
	ipython                   5.1.0                    py35_0
	ipython_genutils          0.1.0                    py35_0
	ipywidgets                5.2.2                    py35_0
	itsdangerous              0.24                     py35_0
	jbig                      2.1                           0
	jdcal                     1.2                      py35_1
	jedi                      0.9.0                    py35_1
	jinja2                    2.8                      py35_1
	jpeg                      8d                            2
	jsonschema                2.5.1                    py35_0
	jupyter                   1.0.0                    py35_3
	jupyter_client            4.4.0                    py35_0
	jupyter_console           5.0.0                    py35_0
	jupyter_core              4.2.0                    py35_0
	jupyterhub                0.7.0                     <pip>
	jupyterhub-ldapauthenticator 1.1                       <pip>
	lazy-object-proxy         1.2.1                    py35_0
	ldap3                     2.1.1                     <pip>
	libdynd                   0.7.2                         0
	libffi                    3.2.1                         0
	libgcc                    4.8.5                         2
	libgfortran               3.0.0                         1
	libpng                    1.6.22                        0
	libsodium                 1.0.10                        0
	libtiff                   4.0.6                         2
	libxcb                    1.12                          0
	libxml2                   2.9.2                         0
	libxslt                   1.1.28                        0
	llvmlite                  0.13.0                   py35_0
	locket                    0.2.0                    py35_1
	lxml                      3.6.4                    py35_0
	Mako                      1.0.6                     <pip>
	markupsafe                0.23                     py35_2
	matplotlib                1.5.3               np111py35_0
	mistune                   0.7.3                    py35_0
	mkl                       11.3.3                        0
	mkl-service               1.1.2                    py35_2
	mpmath                    0.19                     py35_1
	multipledispatch          0.4.8                    py35_0
	nb_anacondacloud          1.2.0                    py35_0
	nb_conda                  2.0.0                    py35_0
	nb_conda_kernels          2.0.0                    py35_0
	nbconvert                 4.2.0                    py35_0
	nbformat                  4.1.0                    py35_0
	nbpresent                 3.0.2                    py35_0
	networkx                  1.11                     py35_0
	nltk                      3.2.1                    py35_0
	nose                      1.3.7                    py35_1
	notebook                  4.2.3                    py35_0
	numba                     0.28.1              np111py35_0
	numexpr                   2.6.1               np111py35_0
	numpy                     1.11.1                   py35_0
	odo                       0.5.0                    py35_1
	openpyxl                  2.3.2                    py35_0
	openssl                   1.0.2j                        0
	pamela                    0.3.0                     <pip>
	pandas                    0.18.1              np111py35_0
	partd                     0.3.6                    py35_0
	patchelf                  0.9                           0
	path.py                   8.2.1                    py35_0
	pathlib2                  2.1.0                    py35_0
	patsy                     0.4.1                    py35_0
	pep8                      1.7.0                    py35_0
	pexpect                   4.0.1                    py35_0
	pickleshare               0.7.4                    py35_0
	pillow                    3.3.1                    py35_0
	pip                       8.1.2                    py35_0
	pixman                    0.32.6                        0
	pkginfo                   1.3.2                    py35_0
	ply                       3.9                      py35_0
	prompt_toolkit            1.0.3                    py35_0
	psutil                    4.3.1                    py35_0
	ptyprocess                0.5.1                    py35_0
	py                        1.4.31                   py35_0
	pyasn1                    0.1.9                    py35_0
	pycosat                   0.6.1                    py35_1
	pycparser                 2.14                     py35_1
	pycrypto                  2.6.1                    py35_4
	pycurl                    7.43.0                   py35_0
	pyflakes                  1.3.0                    py35_0
	pygments                  2.1.3                    py35_0
	pylint                    1.5.4                    py35_1
	pyopenssl                 16.0.0                   py35_0
	pyparsing                 2.1.4                    py35_0
	pyqt                      5.6.0                    py35_0
	pytables                  3.2.3.1             np111py35_0
	pytest                    2.9.2                    py35_0
	python                    3.5.2                         0
	python-dateutil           2.5.3                    py35_0
	python-editor             1.0.3                     <pip>
	pytz                      2016.6.1                 py35_0
	pyyaml                    3.12                     py35_0
	pyzmq                     15.4.0                   py35_0
	qt                        5.6.0                         0
	qtawesome                 0.3.3                    py35_0
	qtconsole                 4.2.1                    py35_1
	qtpy                      1.1.2                    py35_0
	readline                  6.2                           2
	redis                     3.2.0                         0
	redis-py                  2.10.5                   py35_0
	requests                  2.11.1                   py35_0
	rope                      0.9.4                    py35_1
	ruamel_yaml               0.11.14                  py35_0
	scikit-image              0.12.3              np111py35_1
	scikit-learn              0.17.1              np111py35_2
	scipy                     0.18.1              np111py35_0
	setuptools                27.2.0                   py35_0
	simplegeneric             0.8.1                    py35_1
	singledispatch            3.4.0.3                  py35_0
	sip                       4.18                     py35_0
	six                       1.10.0                   py35_0
	snowballstemmer           1.2.1                    py35_0
	sockjs-tornado            1.0.3                    py35_0
	sphinx                    1.4.6                    py35_0
	spyder                    3.0.0                    py35_0
	sqlalchemy                1.0.13                   py35_0
	sqlite                    3.13.0                        0
	statsmodels               0.6.1               np111py35_1
	sudospawner               0.3.0                     <pip>
	sympy                     1.0                      py35_0
	terminado                 0.6                      py35_0
	tk                        8.5.18                        0
	toolz                     0.8.0                    py35_0
	tornado                   4.4.1                    py35_0
	traitlets                 4.3.0                    py35_0
	unicodecsv                0.14.1                   py35_0
	wcwidth                   0.1.7                    py35_0
	werkzeug                  0.11.11                  py35_0
	wheel                     0.29.0                   py35_0
	widgetsnbextension        1.2.6                    py35_0
	wrapt                     1.10.6                   py35_0
	xlrd                      1.0.0                    py35_0
	xlsxwriter                0.9.3                    py35_0
	xlwt                      1.1.2                    py35_0
	xz                        5.2.2                         0
	yaml                      0.1.6                         0
	zeromq                    4.1.4                         0
	zlib                      1.2.8                         3
```
- `jupyter kernelspec list`:
```
Available kernels:
  python3    /opt/anaconda3/lib/python3.5/site-packages/ipykernel/resources
```

**Comment by willingc** [comment]
Thanks for the additional info @mconigliaro.

I did verify that I was able to run R with JupyterHub 0.7 when using pip not a conda environment. I will test conda and your process in the morning. Here's the step that I took for R on JupyterHub 0.7.

1. Install JupyterHub from pip.

2. Installed R from here  http://archive.linux.duke.edu/cran/

3. Installed R studio https://www.rstudio.com/products/rstudio/download/

4. Installed IRkernel from instructions here https://irkernel.github.io and followed steps here https://irkernel.github.io/installation/ for OS X.

5.  Ran `jupyterhub` and was able to select a new R notebook after logging into my JupyterHub account.

Note: `jupyter kernelspec list` run from the command line had this output:

![screen shot 2016-12-08 at 2 18 39 pm](https://cloud.githubusercontent.com/assets/2680980/21030057/71263384-bd51-11e6-89ae-c59e701c8150.png)

I will post again after I get a chance to run using conda.

**Comment by minrk** [comment]
It sounds like you are using Anaconda, and Anaconda's extensions to Jupyter, which give you things like a GUI for installing kernels, and loading kernels from conda envs, which bypass the need to install the kernelspecs in locations that will be found by things like `jupyter kernelspec list`. These aren't part of Jupyter or JupyterHub (I know it's unclear which things come from where). If restarting the notebook server is required to see new kernels, this is probably a question for the [nb_conda_kernels](https://github.com/Anaconda-Platform/nb_conda) and [nb_conda_kernels](https://github.com/Anaconda-Platform/nb_conda_kernels) projects, which provide those things (cc @damianavila).

I'll close #898 and #897 so we can continue this as a single discussion of issues with these extensions, and if there's anything we need to do about interactions between JupyterHub and Anaconda's Jupyter extensions.

**Comment by mconigliaro** [comment]
Yeah, I literally just installed Anaconda, then I followed the instructions in the JupyterHub README. Is there some other doc I should be using?

**Comment by minrk** [comment]
I think R needs a little help.

The first issue is that Anaconda doesn't add the R channel by default, which is reported here: https://github.com/Anaconda-Platform/nb_conda/issues/29

You can add the R channel to your defaults with:

    conda config --add channels r

The second issue may be a known limitation of nb_conda_kernels, I'm not sure. When you say " let you create new notebooks with the new kernel", how does it fail? Is the new kernel not in the list, or does it fail to start somehow? Is refreshing the page sufficient?

**Comment by mconigliaro** [comment]
@minrk So #898 is behavior I've seen after installing every new kernel via the "Conda" tab. You have to log out and log back in before the new kernel will show up in any of the kernel lists. I've never seen R in the kernel lists though, even after logging out and logging back in.

**Comment by damianavila** [comment]
> refreshing the page makes it show up in the kernel lists too

It takes a little bit to catch the new environment and create the corresponding kernelspec because of a caching mechanism: https://github.com/Anaconda-Platform/nb_conda_kernels/blob/master/nb_conda_kernels/manager.py#L42-L65, refreshing do the trick faster...

> I've never seen R in the kernel lists though no matter what I do though.

What do you mean with that? If you try to create an R environment using the conda tab, what experience do you get? What are the kernels do you see? Or you tried to create an R environment from the command line? In that case, what is the conda list for that env?

**Comment by mconigliaro** [comment]
@damianavila #897 explains what happens when I try to create an R kernel in the Conda tab. The only thing I see in the webui is this:

```
An error occurred while creating "testrkernel"
Bad Request
```

But I was able to figure out what was going on by looking at the logs.

**Comment by damianavila** [comment]
> But I was able to figure out what was going on by looking at the logs.

So the missing r channel, I guess...

**Comment by willingc** [comment]
Hi @mconigliaro,

Do you have any objection to me closing this issue? Or are you still having issues with R?

As an FYI, we recently released JupyterHub 0.7.2 as well.
